### ISSUE TITLE
Host Sflow version 2.41-1 Upgrade Issues 

### ISSUE THREAD

**Issue by vidhya-rajan** [sonic-net/sonic-buildimage#13252]
#### Description
host-sflow have few critical fixes for sample flow and SONIC has to be upgraded to the latest version to incorporate those fixes 

#### Steps to reproduce the issue:
Upgrade SONIC to host sflow 2.41-1 latest version 

#### Describe the results you received:
Upgrade will result in compilation issues 
mod_dropmon.c:368:22: error: 'NET_DM_ATTR_HW_DROPS' undeclared (first use in this function); did you mean 'NET_DM_CMD_STOP'?
     attr2.nla_type = NET_DM_ATTR_HW_DROPS;

#### Describe the results you expected:
Compilation of SONIC should be successfull

#### Additional information you deem important (e.g. issue happens only occasionally):
Below files have been modified for successful compilation in order to upgrade host-sflow to latest version

1. rules/config =>  ENABLE_SFLOW_DROPMON = y
The above changes are needed so that compilation issue can be resolved where the workaround header patch for dropmon will be applied. 
2. sflow.mk - HSFLOWD_VERSION = 2.0.41 Updated to upgrade the hsflowd version
3. 0001-sflow-enabled-drop-monitor-support-for-SONiC.patch
This patch has been enhanced to support the latest version since few attributes are removed in latest version 
[Updated-0001-sflow-enabled-drop-monitor-support-for-SONiC.patch.txt](https://github.com/sonic-net/sonic-buildimage/files/10343519/Updated-0001-sflow-enabled-drop-monitor-support-for-SONiC.patch.txt)

Please let us know if the changes are good enough to upgrade hsflowd

**Comment by vidhya-rajan** [comment]
@dgsudharsan @venkatmahalingam @liat-grozovik @hari-selvam
Please let us know if the changes are good enough to upgrade hsflowd

**Comment by dgsudharsan** [comment]
@jeff-yin @padmanarayana Can you please provide feedback?

**Comment by jeff-yin** [comment]
It looks fine to me but @padmanarayana also intends to provide feedback. I was trying to loop in Neil and Peter from InMon, but not sure how I can CC them in Github (username @sflow).

**Comment by sflow** [comment]
@jeff-yin notifications to sflow go to me (Neil).  I'm not sure I understand the problem yet.  It sounds as though the compiling system reports kernel >= 5.4 while its file system is somehow from an older Linux release, so the expected netlink dropmon include-file definitions are not present.  Did I get that right?  And if so, is that a temporary problem?

**Comment by vidhya-rajan** [comment]
Yes your understanding on the problem is right. Build environment (buster) has older kernel version which will not have the required headers . Current SONIC has a local header file patch fix to address this issue .
In order to enable  this patch ENABLE_SFLOW_DROPMON = y has to be set.

**Comment by sflow** [comment]
We checked in a small change to hsflowd and released v2.0.42-1.  The build is no longer conditional on the kernel version.  Instead mod_dropmon.c can now be compiled successfully on older distributions.   However it is not enabled and configured by default.  That is expected to happen in /etc/hsflowd.conf.  This is worth discussing here because the patch above appears to be forcing the dropmon module to be enabled, whether it appears in /etc/hsflowd.conf or not (those changes in hsflowd.c).  Is that what you want for this SONiC profile?  I think it is _probably_ OK to do that.  The module can listen for hw drops, but nothing will happen unless the switch ASIC/driver actually sends to that netlink channel.  We didn't enable the dropmon module by default because not all ASIC/driver implementations report drops yet... but if that is changing soon then maybe we should revisit that?

**Comment by vidhya-rajan** [comment]
Thank you. We forcibly enabled Dropmon module in this patch in order to apply the workaround patch for headers. With the changes in v2.0.42-1 we see that the compilation is successful .  So enabling dropmon by default is no more required and  we can continue to disable by default. Thanks again for the immediate fix.

As highlighted by you not all ASICs may support dropmon so the default can still be retained as disabled. 
However the earlier patch [Updated-0001-sflow-enabled-drop-monitor-support-for-SONiC.patch.txt](https://github.com/sonic-net/sonic-buildimage/files/10343519/Updated-0001-sflow-enabled-drop-monitor-support-for-SONiC.patch.txt) has to be enhanced to be in sync with the latest versions where group attribute of dropmon  is deprecated

**Comment by padmanarayana** [comment]
Thanks @sflow for the v2.0.42-1 release! @vidhya-rajan the plan looks good.

**Comment by bebeisdog** [comment]
Do you have more details regarding the inability to send sample flows for sFlow on Sonic? I have tried everything in the 202211 branch but cannot receive sample flows on the collector.

**Comment by bebeisdog** [comment]
@sflow Thank you for your reply! I have done some troubleshooting.

1. I ensured that my sFlow Docker is enabled (version: 2.0.34) and running, but I only receive counter sample packets.
2. I agree with your statement; I am using the Marvell AC5X. However, I checked the relevant SAI attributes (SAI_SAMPLEPACKET_ATTR_SAMPLE_RATE, SAI_PORT_ATTR_INGRESS_SAMPLEPACKET_ENABLE, etc.), and the chip supports them all.
3. I have also tried this debug mode, and I have traced the general hsflowd process. I am confused about what I should see in debug mode (psample? Netlink event?).

Is it possible that sFlow is not receiving the packets sampled by the ASIC and mirrored to the CPU? If so, this seems to align with the third point you mentioned. But if this is indeed the issue, how can I check it?

Currently, I am thinking of trying to generate packets that do not go through the CPU and then using tcpdump to see if I can capture these packets after a certain period (since sFlow ultimately samples these packets, they should be mirrored to the CPU)."

**Comment by bebeisdog** [comment]
i found my hsflow debug mode got

"
psample netlink (type=28) CMD = 0
"

But then it won't execute the other messages inside the processNetlink_PSAMPLE() function

**Comment by sflow** [comment]
What are the error messages?

Does “sudo lsmod” show that the psample module is loaded in the kernel?  If not you need to run “sudo modprobe psample” (and make it part of the Linux boot sequence”).

And was hsflowd started with root permissions?

**Comment by bebeisdog** [comment]
The psample module is also loaded and started with root privileges. Currently, from the debug messages, it seems that when it runs into the processNetlink_PSAMPLE() function, it receives the following (here I’m pasting the messages related to psample). I ensure that my psample is loaded and that I have obtained the family_id, joined the group, etc. However, after it receives the Netlink Event, the output is as follows:

...
[2025-02-21 16:31:32]  dbg1: psample netlink (type=28) CMD = 0
[2025-02-21 16:31:33]  dbg1: sonic db_select(2)
[2025-02-21 16:31:33]  dbg1: sonic db_select returned 0
[2025-02-21 16:31:33]  dbg1: sonic db_getPortNames()
[2025-02-21 16:31:33]  dbg1: sonic db_getPortNames() returned 0
[2025-02-21 16:31:33]  dbg1: sonic db_select(4)
[2025-02-21 16:31:33]  dbg1: sonic db_select returned 0
[2025-02-21 16:31:33]  dbg1: sonic getLagInfo()
[2025-02-21 16:31:33]  dbg1: sonic getLagInfo() returned 0
[2025-02-21 16:31:33]  dbg1: sonic db_selectCB: reply=status(0)="OK"
[2025-02-21 16:31:33]  dbg1: sonic db_portNamesCB: reply=array(108)
[2025-02-21 16:31:33]  dbg1: sonic db_selectCB: reply=status(0)="OK"
[2025-02-21 16:31:33]  dbg1: sonic getLagInfoCB: reply=array(0)
[2025-02-21 16:31:33]  dbg1: pollCounters(adaptor=Ethernet53)
...

It seems that it is not processing this event further, but I don't see any other error messages inside.
Because if flow samples can be emitted, we should then see this line of debug message:

myDebug(2, "psample: grp=%u in=%u out=%u n=%u seq=%u drops=%u pktlen=%u",
        grp_no,
        ifin,
        ifout,
        sample_n,
        grp_seq,
        drops,
        pkt_len);

**Comment by sflow** [comment]
I don't see any error messages here.  It just looks like nothing was posted to the psample netlink channel.

As a crosscheck, I suggest you run "psampletest" on the switch at the same time:

https://github.com/sflow/psampletest

It sends a dummy sample (which hsflowd should see and log),  then listens and prints whatever else might appear.

**Comment by bebeisdog** [comment]
@sflow  Thx, Thank you very much for your help. I also captured the Netlink events, and I think I can categorize the issue as related to the behavior of Chip & SAI. :)
